src/cargo/core/gc.rs

```diff
@@ -0,0 +1,557 @@
+//! Support for garbage collecting unused files from downloaded files or
+//! artifacts from the target directory.
+//!
+//! The [`Gc`] type provides the high-level interface for the
+//! garbage-collection system.
+//!
+//! Garbage collection can be done "automatically" by cargo, which it does by
+//! default once a day when running any command that does a lot of work (like
+//! `cargo build`). The entry point for this is the [`auto_gc`] function,
+//! which handles some basic setup, creating the [`Gc`], and calling
+//! [`Gc::auto`].
+//!
+//! Garbage collection can also be done manually via the `cargo clean` command
+//! by passing any option that requests deleting unused files. That is
+//! implemented by calling the [`Gc::gc`] method.
+//!
+//! Garbage collection for the global cache is guided by the last-use tracking
+//! implemented in the [`crate::core::global_cache_tracker`] module. See that
+//! module documentation for an in-depth explanation of how global cache
+//! tracking works.
+
+use crate::core::global_cache_tracker::{self, GlobalCacheTracker};
+use crate::ops::CleanContext;
+use crate::util::cache_lock::{CacheLock, CacheLockMode};
+use crate::{CargoResult, Config};
+use anyhow::format_err;
+use anyhow::{bail, Context};
+use serde::Deserialize;
+use std::time::Duration;
+
+/// Performs automatic garbage collection.
+///
+/// This is called in various places in Cargo where garbage collection should
+/// be performed automatically based on the config settings. The default
+/// behavior is to only clean once a day.
+///
+/// This should only be called in code paths for commands that are already
+/// doing a lot of work. It should only be called *after* crates are
+/// downloaded so that the last-use data is updated first.
+///
+/// It should be cheap to call this multiple times (subsequent calls are
+/// ignored), but try not to abuse that.
+pub fn auto_gc(config: &Config) {
+    if !config.cli_unstable().gc {
+        return;
+    }
+    if !config.network_allowed() {
+        // As a conservative choice, auto-gc is disabled when offline. If the
+        // user is indefinitely offline, we don't want to delete things they
+        // may later depend on.
+        tracing::trace!("auto gc disabled, running offline");
```

Comment:
nit: compared to how the other traces are structured, this should be `"running offline, auto gc disabled"`